### PR TITLE
Add __call__ method to Link for backward compatibility

### PR DESCRIPTION
**Pull request by mitmul**
This PR adds `__call__` to Link. This is for backporting to v4, so once this is merged, the `__call__` method added to Link in v5 will be overwritten by #4735.

### CODE REVIEW

**Comment by niboshi** [comment]
How about something like this? It seems actually faster.

```
try:
    forward = self.forward
except AttributeError:
    raise ...
forward(...)
```

**Comment by mitmul** [comment]
@niboshi Thanks. Fixed

**Comment by niboshi** [comment]
Please fix test failure.

File: chainer/link.py [hunk 1]
```diff
@@ -194,6 +194,15 @@ def __init__(self):
         finally:
             self._within_init_scope = old_flag
 
+    def __call__(self, *args, **kwargs):
+        try:
+            forward = self.forward
+        except AttributeError:
+            raise TypeError(
+                    '{} object does not have neither \'__call__\' nor '
+                    '\'forward\' method.'.format(self))
```
Comment (niboshi) [inline chainer/link.py]: This sentence is not accurate (`self` has `__call__`).

"Neither Link.\_\_call__() method is overridden nor forward() method is defined."?

Comment (mitmul) [reply chainer/link.py]: Thanbks, fixed

File: chainer/link.py [hunk 2]
```diff
@@ -194,6 +194,15 @@ def __init__(self):
         finally:
             self._within_init_scope = old_flag
 
+    def __call__(self, *args, **kwargs):
+        try:
+            forward = self.forward
+        except AttributeError:
+            raise TypeError(
+                    '{} object does not have neither a method which overrides
+                    '\'Link.__call__\' nor a \'forward\' method.'.format(self))
```
Comment (niboshi) [inline chainer/link.py]: Hmm, I don't think this sentence is grammatically correct.

"{} object has neither `Link.__call__` method overridden nor `forward` method defined"?

Comment (mitmul) [reply chainer/link.py]: oops, right. I'll fix it

**Comment by mitmul** [comment]
@niboshi Sorry fixed

File: tests/chainer_tests/test_link.py [hunk 3]
```diff
@@ -1751,4 +1751,36 @@ def test_gpu_to_intel64_unsupported(self):
...
+            def __init__(self):
+                super(Model, self).__init__()
+
+        self.model = Model()
+
+    def test_has_forward_no_call(self):
+        self.model.forward = mock.MagicMock()
+        self.model(0)  # Link.__call__ is called
+        self.model.forward.assert_called_once()
+        self.assertTrue(self.model._found_forward)
```
Comment (niboshi) [inline tests/chainer_tests/test_link.py]: Please fix this test.
>self.model(0)  # Link.__call__ is called

-> "model.forward is called"

> self.assertTrue(self.model._found_forward)

Remove this line

**Comment by niboshi** [comment]
Please fix style errors, too.

**Comment by mitmul** [comment]
@niboshi Thanks, fixed

**Comment by kmaehashi** [comment]
CIFAR10 example in master branch is broken due to PR #4735. Is this PR ready for merge?

File: chainer/link.py [hunk 4]
```diff
@@ -197,6 +197,15 @@ def __init__(self):
         finally:
             self._within_init_scope = old_flag
 
+    def __call__(self, *args, **kwargs):
+        try:
+            forward = self.forward
+        except AttributeError:
+            raise TypeError(
+                '{} object has neither a \'Link.__call__\' method overridden'
+                'nor a forward method defined.'.format(self))
```
Comment (niboshi) [inline chainer/link.py]: Both articles `a`  are not needed.
"forward" should be surrounded by `''` like `Link.__call__`.

**Comment by niboshi** [comment]
@kmaehashi How about reverting #4735?

**Comment by okuta** [comment]
Sorry, that is my mistake.

File: chainer/link.py (hunk 4 above)
Comment (mitmul) [reply chainer/link.py]: Oooops thanks 😉

**Comment by mitmul** [comment]
@niboshi Fixed

File: chainer/link.py [hunk 5]
```diff
@@ -197,6 +197,15 @@ def __init__(self):
         finally:
             self._within_init_scope = old_flag
 
+    def __call__(self, *args, **kwargs):
+        try:
+            forward = self.forward
+        except AttributeError:
+            raise TypeError(
+                '{} object has neither \'Link.__call__\' method overridden'
+                'nor \'forward\' method defined.'.format(self))
```
Comment (niboshi) [inline chainer/link.py]: Missing space between "overridden" and "nor"...

Comment (mitmul) [reply chainer/link.py]: Oops! Thanks!

**Comment by niboshi** [comment]
LGTM!

**Pull request merged**